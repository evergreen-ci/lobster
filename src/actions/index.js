// @flow strict

import type { Action as LogviewerAction } from './logviewer';
import type { Log } from '../models';

export const LOGKEEPER_LOAD_DATA = 'logkeeper:load-data';
export const LOBSTER_LOAD_DATA = 'lobster:load-data';
export const PROCESS_RESPONSE = 'process-response';
export const LOAD_CACHED_DATA = 'load-cached-data';
export const EVERGREEN_LOAD_DATA = 'evergreen:load-data';
export const SETUP_CACHE = 'setup-cache';
export const WIPE_CACHE = 'wipe-cache';

export type LogkeeperLoadData = {|
  +type: 'logkeeper:load-data',
  +payload: {|
    +build: string,
    +test: ?string
  |}
|}

export type LobsterLoadData = {|
  +type: 'lobster:load-data',
  +payload: {|
    +url: string,
    +server: string
  |}
|}

const logTypes_ = {
  'resmoke': '',
  'raw': ''
};

export const logTypes = (): string[] => Object.keys(logTypes_);

export type LogType = $Keys<typeof logTypes_>

export type ProcessResponse = {|
  +type: 'process-response',
  +payload: {|
    +type: LogType,
    +data: string,
    +isDone: boolean
  |},
  +error: boolean
|}

export function logkeeperLoadData(build: string, test: ?string): LogkeeperLoadData {
  return {
    type: LOGKEEPER_LOAD_DATA,
    payload: {
      build: build,
      test: test
    }
  };
}

// Load data from the lobster server
export function lobsterLoadData(server: string, url: string): LobsterLoadData {
  return {
    type: LOBSTER_LOAD_DATA,
    payload: {
      url: url,
      server: server
    }
  };
}

export function processData(data: string, type: LogType, isDone?: boolean): ProcessResponse {
  return {
    type: PROCESS_RESPONSE,
    payload: {
      type: type,
      data: data,
      isDone: isDone || false
    },
    error: false
  };
}

export function processLocalData(data: string, type: LogType): ProcessResponse {
  return {
    type: PROCESS_RESPONSE,
    payload: {
      type: type,
      data: data,
      isDone: true
    },
    error: false
  };
}

export function processDataError(data: string): ProcessResponse {
  return {
    type: PROCESS_RESPONSE,
    payload: {
      type: 'resmoke',
      data: data,
      isDone: true
    },
    error: true
  };
}

export type LoadCachedData = {|
  type: 'load-cached-data',
  payload: {|
    log: Log
  |}
|}

export function loadCachedData(data: Log): LoadCachedData {
  return {
    type: LOAD_CACHED_DATA,
    payload: {
      log: data
    }
  };
}

export type CacheStatus = 'ok' | 'error' | 'never' | 'later' | 'unsupported' | null;

export type SetupCache = {|
  +type: 'setup-cache',
  +payload: {
    +status: CacheStatus,
    +size: number
  },
  +error: boolean
|}

export function setCache(status: CacheStatus, size: number): SetupCache {
  return {
    type: SETUP_CACHE,
    payload: {
      status: status,
      size: size
    },
    error: false
  };
}

const evergreenTaskLogTypes: { [string]: string } = {
  'all': 'ALL',
  'task': 'T',
  'agent': 'A',
  'system': 'S'
  // 'event': 'E' // Not actually supported by the api
};

export type EvergreenTaskLogType = $Keys<typeof evergreenTaskLogTypes>;

export function stringToInteralEvergreenTaskLogType(a: string): ?string {
  return evergreenTaskLogTypes[a];
}

export function stringToEvergreenTaskLogType(a: string): ?EvergreenTaskLogType {
  if (!evergreenTaskLogTypes[a]) {
    return null;
  }

  return a;
}

export type EvergreenTaskLog = {|
  type: 'task',
  id: string,
  execution: number,
  log: EvergreenTaskLogType
|};

export type EvergreenTestLog = {|
  type: 'test',
  id: string,
|};

export type EvergreenLoadData = {|
  +type: 'evergreen:load-data',
  +payload: EvergreenTaskLog | EvergreenTestLog
|}

export function evergreenLoadTaskLog(id: string, execution: number, log: EvergreenTaskLogType): EvergreenLoadData {
  return {
    type: 'evergreen:load-data',
    payload: {
      type: 'task',
      id: id,
      execution: execution,
      log: log
    }
  };
}

export function evergreenLoadTestLog(id: string): EvergreenLoadData {
  return {
    type: 'evergreen:load-data',
    payload: {
      type: 'test',
      id: id
    }
  };
}

export type WipeCache = {|
  type: 'wipe-cache',
  payload: {|
    file: ?string
  |}
|}

export function wipeCache(): WipeCache {
  return {
    type: WIPE_CACHE,
    payload: {
      file: null
    }
  };
}

export function fromFileFromCache(f: string): WipeCache {
  return {
    type: WIPE_CACHE,
    payload: {
      file: f
    }
  };
}
export type Action = ProcessResponse
  | LoadCachedData
  | WipeCache
  | LogkeeperLoadData
  | LobsterLoadData
  | EvergreenLoadData
  | LogviewerAction
