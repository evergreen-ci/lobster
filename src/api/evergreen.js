// @flow strict

import type { EvergreenLog, EvergreenTaskLogType } from '../models';
import { EVERGREEN_BASE } from '../config';
import { stringToInteralEvergreenTaskLogType } from '../models';

export function taskLogURL(id: string, execution: number, type: EvergreenTaskLogType): string {
  const logType = stringToInteralEvergreenTaskLogType(type) || 'ALL';
  return `${EVERGREEN_BASE}/task_log_raw/${id}/${execution}?type=${logType}`;
}

export function taskLogRawURL(id: string, execution: number, type: EvergreenTaskLogType): string {
  return `${taskLogURL(id, execution, type)}&text=true`;
}

export function testLogURL(id: string): string {
  return `${EVERGREEN_BASE}/test_log/${id}`;
}

export function testLogRawURL(id: string): string {
  return `${testLogURL(id)}?raw=1`;
}

export function taskURL(taskID: string, execution: ?number): string {
  const base = `${EVERGREEN_BASE}/task/${taskID}`;

  if (execution == null) {
    return base;
  }

  return base + `/${execution}`;
}

export function fetchEvergreen(log: EvergreenLog): Promise<Response> {
  const init = { method: 'GET', credentials: 'include' };
  let req;
  if (log.type === 'evergreen-task') {
    req = new Request(taskLogRawURL(log.id, log.execution, log.log), init);
  } else if (log.type === 'evergreen-test') {
    req = new Request(testLogRawURL(log.id), init);
  }

  return window.fetch(req);
}
