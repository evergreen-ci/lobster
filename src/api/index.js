// @flow strict

export * from './logkeeper';
export * from './evergreen';
