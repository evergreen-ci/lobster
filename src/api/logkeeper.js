// @flow strict

import { LOGKEEPER_BASE } from '../config';

function generateLogkeeperUrl(buildParam: string, testParam: ?string): string {
  if (!buildParam) {
    return '';
  }
  if (!testParam) {
    return LOGKEEPER_BASE + '/build/' + buildParam + '/all?raw=1';
  }
  return LOGKEEPER_BASE + '/build/' + buildParam + '/test/' + testParam + '?raw=1';
}

export function fetchLogkeeper(build: string, test: ?string): Promise<Response> {
  const req = new Request(generateLogkeeperUrl(build, test), { method: 'GET' });
  return window.fetch(req);
}

export function fetchLobster(server: string, url: string): Promise<Response> {
  const init = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      url: url
    })
  };
  const req = new Request(`http://${server}`, init);
  return window.fetch(req);
}
