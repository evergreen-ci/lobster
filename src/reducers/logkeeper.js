// @flow strict

import type { Action, Log } from '../actions';
import { LOGKEEPER_LOAD_RESPONSE } from '../actions';
import * as LogProcessor from './LogProcessor';

const initialState: Log = {
  lines: [],
  colorMap: new Map()
};

export function logkeeperDataResponse(state: Log = initialState, action: Action): Log {
  if (action.type !== LOGKEEPER_LOAD_RESPONSE || action.error) {
    return state;
  }

  switch (action.payload.type) {
    case 'resmoke':
      return LogProcessor.resmoke(action.payload.data);

    case 'raw':
      return LogProcessor.raw('\n')(action.payload.data);

    default:
      return initialState;
  }
}
