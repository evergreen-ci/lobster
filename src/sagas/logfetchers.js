// @flow strict

import { put, call, takeEvery } from 'redux-saga/effects';
import type { Saga } from 'redux-saga';
import * as actions from '../actions';
import * as api from '../api/logkeeper';
// import cachedRequest from '../api/cachedRequest';
import { fetchEvergreen } from '../api/evergreen';
import { writeToCache, readFromCache } from './lobstercage';

export function* logkeeperLoadData(action: actions.LogkeeperLoadData): Saga<void> {
  console.log('fetch (logkeeper)', action.payload.build, action.payload.test);
  const test = action.payload.test || 'all';
  const f = `fetchLogkeeper-${action.payload.build}-${test}.json`;
  try {
    try {
      yield call(readFromCache, f);
    } catch (_err) {
      const resp = yield call(api.fetchLogkeeper, action.payload.build, action.payload.test);
      if (resp.status !== 200) {
        throw resp;
      }

      const data = yield resp.text();
      yield put.resolve(actions.processData(data, 'resmoke', true));
      yield call(writeToCache, f);
    }
  } catch (error) {
    yield put(actions.processDataError(error));
  }
}

export function* lobsterLoadData(action: actions.LobsterLoadData): Saga<void> {
  console.log('fetch (lobster server)', action.payload.server, action.payload.url);
  try {
    const resp = yield call(api.fetchLobster, action.payload.server, action.payload.url);
    if (resp.status !== 200) {
      throw resp;
    }

    const data = yield resp.text();
    yield put(actions.processData(data, 'resmoke', true));
  } catch (error) {
    yield put(actions.processDataError(error));
  }
}

export function* evergreenLoadData(action: actions.EvergreenLoadData): Saga<void> {
  console.log(`fetch (evergreen) ${JSON.stringify(action.payload)}`);
  try {
    const resp = yield call(fetchEvergreen, action.payload);
    if (resp.status !== 200) {
      throw resp;
    }

    const body = yield resp.text();
    yield put(actions.processData(body, 'raw', true));
  } catch (error) {
    yield put(actions.processDataError(error));
  }
}

export default function* fetcher(): Saga<void> {
  yield takeEvery(actions.LOBSTER_LOAD_DATA, lobsterLoadData);
  yield takeEvery(actions.LOGKEEPER_LOAD_DATA, logkeeperLoadData);
  yield takeEvery(actions.EVERGREEN_LOAD_DATA, evergreenLoadData);
}
