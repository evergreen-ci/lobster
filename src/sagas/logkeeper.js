// @flow strict

import { put, call } from 'redux-saga/effects';
import type { Saga } from 'redux-saga';
import * as actions from '../actions';
import * as api from '../api/logkeeper';

export function* logkeeperLoadData(action: actions.LogkeeperLoadData): Saga<void> {
  console.log('fetch', action.payload.build, action.payload.test);
  try {
    const data = yield call(api.fetchLogkeeper, action.payload.build, action.payload.test);
    yield put(actions.logkeeperDataSuccess(data.data));
  } catch (error) {
    yield put(actions.logkeeperDataError(error));
  }
}

export function* lobsterLoadData(action: actions.LobsterLoadData): Saga<void> {
  console.log('fetch', action.payload.server, action.payload.url);
  try {
    const data = yield call(api.fetchLobster, action.payload.server, action.payload.url);
    yield put(actions.logkeeperDataSuccess(data.data));
  } catch (error) {
    yield put(actions.logkeeperDataError(error));
  }
}

