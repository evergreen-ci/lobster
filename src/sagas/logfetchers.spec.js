import sinon from 'sinon';
import assert from 'assert';
import { runSaga } from 'redux-saga';
import * as api from '../api/logkeeper';
import * as actions from '../actions';
import * as sagas from './logfetchers';

describe('lobsterLoadData', function() {
  afterEach(() => sinon.restore());

  test('resolve', function() {
    sinon.replace(api, 'fetchLobster', sinon.stub().resolves(new Response('lobster')));

    const action = actions.lobsterLoadData('localhost:9001', 'a.log');

    const dispatch = sinon.fake();
    const getState = sinon.fake.returns({
      cache: {
        size: 123
      }
    });
    const options = {
      dispatch: dispatch,
      getState: getState
    };
    const result = runSaga(options, sagas.lobsterLoadData, action).done;
    assert.ok(result);

    result.then(function() {
      assert.strictEqual(api.fetchLobster.callCount, 1);
      assert.deepEqual(api.fetchLobster.firstCall.args, ['localhost:9001', 'a.log']);

      assert.strictEqual(dispatch.callCount, 1);
      assert.strictEqual(dispatch.firstCall.args.length, 1);
      assert.strictEqual(dispatch.firstCall.args[0].type, actions.PROCESS_RESPONSE);
      assert.strictEqual(dispatch.firstCall.args[0].payload.data, 'lobster');
      assert.strictEqual(dispatch.firstCall.args[0].error, false);
      assert.strictEqual(dispatch.lastCall.args[0].payload.isDone, true);
    });
  });

  test('fail', function() {
    sinon.replace(api, 'fetchLobster', sinon.stub().rejects('error'));

    const action = actions.lobsterLoadData('localhost:9001', 'a.log');
    const dispatch = sinon.fake();
    const result = runSaga({ dispatch: dispatch }, sagas.lobsterLoadData, action).done;
    assert.ok(result);
    result.then(function() {
      assert.strictEqual(api.fetchLobster.callCount, 1);
      assert.deepEqual(api.fetchLobster.firstCall.args, ['localhost:9001', 'a.log']);

      assert.strictEqual(dispatch.callCount, 1);
      assert.strictEqual(dispatch.firstCall.args[0].type, actions.PROCESS_RESPONSE);
      assert.strictEqual(dispatch.firstCall.args[0].error, true);
    });
  });
});

describe('logkeeperLoadData', function() {
  afterEach(() => sinon.restore());

  test('resolve', function() {
    sinon.replace(api, 'fetchLogkeeper', sinon.stub().resolves(new Response('logkeeper')));

    const action = actions.logkeeperLoadData('build0', 'test0');
    const dispatch = sinon.fake();
    const getState = sinon.fake.returns({
      cache: {
        size: 123
      }
    });
    const result = runSaga({ dispatch: dispatch, getState: getState }, sagas.logkeeperLoadData, action).done;
    assert.ok(result);

    result.then(function() {
      assert.strictEqual(api.fetchLogkeeper.callCount, 1);
      assert.deepEqual(api.fetchLogkeeper.firstCall.args, ['build0', 'test0']);

      assert.strictEqual(dispatch.callCount, 1);
      assert.strictEqual(dispatch.firstCall.args.length, 1);
      assert.strictEqual(dispatch.firstCall.args[0].type, actions.PROCESS_RESPONSE);
      assert.strictEqual(dispatch.firstCall.args[0].payload.data, 'logkeeper');
      assert.strictEqual(dispatch.firstCall.args[0].error, false);
      assert.strictEqual(dispatch.lastCall.args[0].payload.isDone, true);
    });
  });

  test('resolve-error', function() {
    sinon.replace(api, 'fetchLogkeeper', sinon.stub().resolves(new Response('nope logkeeper', { status: 500 })));

    const action = actions.logkeeperLoadData('build0');
    const dispatch = sinon.fake();
    const getState = sinon.fake.returns({
      cache: {
        size: 123
      }
    });
    const result = runSaga({ dispatch: dispatch, getState: getState }, sagas.logkeeperLoadData, action).done;
    assert.ok(result);

    result.then(function() {
      assert.strictEqual(api.fetchLogkeeper.callCount, 1);
      assert.deepEqual(api.fetchLogkeeper.firstCall.args, ['build0', undefined]);

      assert.strictEqual(dispatch.callCount, 1);
      assert.strictEqual(dispatch.firstCall.args.length, 1);
      assert.strictEqual(dispatch.firstCall.args[0].type, actions.PROCESS_RESPONSE);

      dispatch.firstCall.args[0].payload.data.text().then(function(text) {
        assert.strictEqual(text, 'nope logkeeper');
      });
      assert.strictEqual(dispatch.firstCall.args[0].error, true);
    });
  });

  test('reject', function() {
    sinon.replace(api, 'fetchLogkeeper', sinon.stub().rejects('error'));

    const action = actions.logkeeperLoadData('build0', 'test0');
    const dispatch = sinon.fake();
    const getState = sinon.fake.returns({
      cache: {
        size: 123
      }
    });
    const result = runSaga({ dispatch: dispatch, getState: getState }, sagas.logkeeperLoadData, action).done;
    assert.ok(result);

    result.then(function() {
      assert.strictEqual(api.fetchLogkeeper.callCount, 1);
      assert.deepEqual(api.fetchLogkeeper.firstCall.args, ['build0', 'test0']);

      assert.strictEqual(dispatch.callCount, 1);
      assert.strictEqual(dispatch.firstCall.args.length, 1);
      assert.strictEqual(dispatch.firstCall.args[0].type, actions.PROCESS_RESPONSE);
      assert.strictEqual(dispatch.firstCall.args[0].error, true);
    });
  });
});
