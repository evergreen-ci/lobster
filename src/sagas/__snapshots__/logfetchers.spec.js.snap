// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchEvergreen resolves-200-task 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "execution": 0,
            "id": "task",
            "log": "all",
            "type": "evergreen-task",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/evergreenLoadData",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "execution": 0,
            "id": "task",
            "log": "all",
            "type": "evergreen-task",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/proxy",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "payload": Object {
            "data": "evergreen-task",
            "isDone": true,
            "type": "raw",
          },
          "type": "process-response",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`fetchEvergreen resolves-200-test 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "id": "test",
            "type": "evergreen-test",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/evergreenLoadData",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "fetchEvergreen-test-test.json",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/readFromCache",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "id": "test",
            "type": "evergreen-test",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/proxy",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "fetchEvergreen-test-test.json",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/writeToCache",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "payload": Object {
            "data": "evergreen-test",
            "isDone": true,
            "type": "raw",
          },
          "type": "process-response",
        },
        "channel": null,
        "resolve": true,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
}
`;

exports[`fetchLobster rejects 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "file": "simple.log",
            "server": "domain.invalid",
            "type": "lobster",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/lobsterLoadData",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "domain.invalid",
          "simple.log",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/proxy",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "payload": Object {
            "data": Object {
              "name": "error",
            },
            "isDone": true,
            "type": "resmoke",
          },
          "type": "process-response",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`fetchLobster resolves-200 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "file": "simple.log",
            "server": "domain.invalid",
            "type": "lobster",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/lobsterLoadData",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "domain.invalid",
          "simple.log",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/proxy",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "payload": Object {
            "data": "lobster",
            "isDone": true,
            "type": "resmoke",
          },
          "type": "process-response",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`fetchLobster resolves-404 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "file": "simple.log",
            "server": "domain.invalid",
            "type": "lobster",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/lobsterLoadData",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "domain.invalid",
          "simple.log",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/proxy",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": true,
          "payload": Object {
            "data": Object {
              "_bodyBlob": Object {},
              "_bodyInit": Object {},
              "headers": Object {
                "map": Object {},
              },
              "ok": false,
              "status": 404,
              "statusText": "OK",
              "type": "default",
              "url": "",
            },
            "isDone": true,
            "type": "resmoke",
          },
          "type": "process-response",
        },
        "channel": null,
      },
    },
  ],
}
`;

exports[`fetchLogkeeper resolves-200 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "build": "build",
            "test": "test",
            "type": "logkeeper",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/logkeeperLoadData",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "fetchLogkeeper-build-test.json",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/readFromCache",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "build",
          "test",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/proxy",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "fetchLogkeeper-build-test.json",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/writeToCache",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "payload": Object {
            "data": "logkeeper",
            "isDone": true,
            "type": "resmoke",
          },
          "type": "process-response",
        },
        "channel": null,
        "resolve": true,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
}
`;

exports[`fetchLogkeeper resolves-200-notest 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {
            "build": "build",
            "type": "logkeeper",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/logkeeperLoadData",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "fetchLogkeeper-build-all.json",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/readFromCache",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "build",
          undefined,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/proxy",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "fetchLogkeeper-build-all.json",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/writeToCache",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "error": false,
          "payload": Object {
            "data": "logkeeper",
            "isDone": true,
            "type": "resmoke",
          },
          "type": "process-response",
        },
        "channel": null,
        "resolve": true,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
}
`;
